crates/miden-tx/src/executor/exec_host.rs

```diff
@@ -362,6 +364,106 @@ where
         Ok(asset_witness_to_advice_mutation(asset_witness))
     }
 
+    /// Handles a request for a note script by querying the data store.
+    ///
+    /// The script is fetched from the data store, and then the NoteRecipient is built and used
+    /// to create the OutputNoteBuilder.
+    async fn on_note_script_requested(
+        &mut self,
+        script_root: Word,
+        metadata: miden_objects::note::NoteMetadata,
+        recipient_digest: Word,
+        note_idx: usize,
+        process: &ProcessState<'_>,
+    ) -> Result<Vec<AdviceMutation>, TransactionKernelError> {
+        // Try to get the note script from the data store
+        let note_script_result = self.base_host.store().get_note_script(script_root).await;
+
+        let (recipient, mutations) = match note_script_result {
+            Ok(note_script) => {
+                // Script found - build the recipient
+                let script_felts: Vec<Felt> = (&note_script).into();
+
+                // Get recipient data from advice provider
+                let data = process
+                    .advice_provider()
+                    .get_mapped_values(&recipient_digest)
+                    .ok_or_else(|| {
+                        TransactionKernelError::other("recipient data should be in advice provider")
+                    })?;
+
+                if data.len() != 13 {
+                    return Err(TransactionKernelError::MalformedRecipientData(data.to_vec()));
+                }
+
+                // NoteDetails from AdviceProvider: [num_inputs, INPUTS_COMMITMENT, SCRIPT_ROOT,
+                // SERIAL_NUM]
+                let num_inputs = data[0].as_int() as usize;
+                let inputs_commitment = Word::new([data[1], data[2], data[3], data[4]]);
+                let serial_num = Word::from([data[9], data[10], data[11], data[12]]);
+
+                let inputs_data = process.advice_provider().get_mapped_values(&inputs_commitment);
+
+                let inputs = match inputs_data {
+                    None => NoteInputs::default(),
+                    Some(inputs) => {
```

Comment:
Similar code is used in another place as well. Could we create a helper function that encapsulates this logic? It could look something like:
```rust
fn read_recipient_info(
    advice: &AdviceProvider,
    recipient_digest: Word,
) -> Result<(NoteInputs, Word, Word), TransactionKernelError> {
    // return (NoteInputs, script_root, serial_num)
}
```